hazelcast-wm/src/main/java/com/hazelcast/web/HazelcastHttpSession.java

```diff
@@ -0,0 +1,368 @@
+        if (cacheEntry == null || (cacheEntry.isReload() && !cacheEntry.isDirty())) {
+            Object value = null;
+            if (!transientAttribute) {
+                try {
+                    value = webFilter.getClusteredSessionService().getAttribute(id, name);
+                } catch (Exception ignored) {
+                    EmptyStatement.ignore(ignored);
+                }
+            }
+            if (value == null) {
+                cacheEntry = WebFilter.NULL_ENTRY;
+            } else {
+                cacheEntry = new LocalCacheEntry(transientAttribute);
+                cacheEntry.setValue(value);
+                cacheEntry.setReload(false);
+            }
+            localCache.put(name, cacheEntry);
+        }
+        return cacheEntry != WebFilter.NULL_ENTRY ? cacheEntry.getValue() : null;
+    }
+
+    private Object getLocalAttribute(String name) {
+        LocalCacheEntry cacheEntry = localCache.get(name);
+        if (cacheEntry == null || cacheEntry.isRemoved()) {
+            return null;
+        }
+        return cacheEntry.getValue();
+    }
+
+    public Enumeration<String> getAttributeNames() {
+        final Set<String> keys = selectKeys();
+        return new Enumeration<String>() {
+            private final String[] elements = keys.toArray(new String[keys.size()]);
+            private int index;
+
+            @Override
+            public boolean hasMoreElements() {
+                return index < elements.length;
+            }
+
+            @Override
+            public String nextElement() {
+                return elements[index++];
+            }
+        };
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public ServletContext getServletContext() {
+        return webFilter.servletContext;
+    }
+
+    @Deprecated
+    @SuppressWarnings("deprecation")
+    public HttpSessionContext getSessionContext() {
+        return originalSession.getSessionContext();
+    }
+
+    public Object getValue(final String name) {
+        return getAttribute(name);
+    }
+
+    public String[] getValueNames() {
+        final Set<String> keys = selectKeys();
+        return keys.toArray(new String[keys.size()]);
+    }
+
+    public void invalidate() {
+        // we must invalidate hazelcast session first
+        // invalidating original session will trigger another
+        // invalidation as our SessionListener will be triggered.
+        webFilter.destroySession(this, true);
+        originalSession.invalidate();
+    }
+
+    public boolean isNew() {
+        return originalSession.isNew() && clusterWideNew;
+    }
+
+    public void putValue(final String name, final Object value) {
+        setAttribute(name, value);
+    }
+
+    public void removeAttribute(final String name) {
+        LocalCacheEntry entry = localCache.get(name);
+        if (entry != null && entry != WebFilter.NULL_ENTRY) {
+            entry.setValue(null);
+            entry.setRemoved(true);
```

Comment:
`LocalCacheEntry.removed` is used only here inside `getAttribute `  https://github.com/hazelcast/hazelcast/pull/5373/files#diff-d770d63698a9e0e449d0d8ad43f24943R120 even if you set it to false it will return null. So lets leave this with true for now.